core/arch/riscv64/mangle.c

```diff
@@ -32,40 +32,170 @@
 
 #include "../globals.h"
 #include "arch.h"
+#include "instr_create_shared.h"
+#include "instrument.h" /* instrlist_meta_preinsert */
+
+/* Make code more readable by shortening long lines.
+ * We mark everything we add as non-app instr.
+ */
+#define PRE instrlist_meta_preinsert
 
 void
 mangle_arch_init(void)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    /* Nothing. */
 }
 
 void
 insert_clear_eflags(dcontext_t *dcontext, clean_call_info_t *cci, instrlist_t *ilist,
                     instr_t *instr)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    /* Nothing. */
 }
 
 uint
 insert_push_all_registers(dcontext_t *dcontext, clean_call_info_t *cci,
                           instrlist_t *ilist, instr_t *instr, uint alignment,
-                          opnd_t push_pc,
-                          reg_id_t scratch /*optional*/
-                              _IF_AARCH64(bool out_of_line))
+                          opnd_t push_pc, reg_id_t scratch)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
-    return 0;
+    uint dstack_offs = 0;
+    int max_offs;
+
+    if (cci == NULL)
+        cci = &default_clean_call_info;
+    ASSERT(proc_num_simd_registers() == MCXT_NUM_SIMD_SLOTS);
+
+    /* a0 is used to save and restore the pc and csr registers. */
+    cci->reg_skip[DR_REG_A0 - DR_REG_START_GPR] = false;
+
+    max_offs = get_clean_call_switch_stack_size();
+
+    PRE(ilist, instr,
+        INSTR_CREATE_addi(dcontext, opnd_create_reg(DR_REG_SP),
+                          opnd_create_reg(DR_REG_SP),
+                          opnd_add_flags(opnd_create_immed_int(-max_offs, OPSZ_12b),
+                                         DR_OPND_IMM_PRINT_DECIMAL)));
+
+    /* Push GPRs. */
+    for (int i = 1; i < DR_NUM_GPR_REGS; i += 1) {
+        if (cci->reg_skip[i])
+            continue;
+
+        PRE(ilist, instr,
+            INSTR_CREATE_sd(dcontext,
+                            opnd_add_flags(opnd_create_base_disp(
+                                               DR_REG_SP, DR_REG_NULL, 0,
+                                               dstack_offs + i * sizeof(reg_t), OPSZ_8),
```

Comment:
`sizeof(reg_t)` here but `XSP_SZ` on line 93: use the same one for both (I guess `XSP_SZ` is what the code on other arches is using)